apps/spruce/src/gql/client/cache.ts

```diff
@@ -17,6 +18,144 @@ export const cache = new InMemoryCache({
         hasVersion: {
           keyArgs: ["$patchId"],
         },
+        waterfall: {
+          keyArgs: ["options", ["projectIdentifier"]],
+          read(existing, { args, readField, variables }) {
+            if (!existing) {
+              return undefined;
+            }
+
+            const minOrder = args?.options?.minOrder ?? 0;
+            const maxOrder = args?.options?.maxOrder ?? 0;
+            const limit = variables?.options?.limit ?? 0;
+
+            const existingVersions =
+              readField<WaterfallQuery["waterfall"]["flattenedVersions"]>(
+                "flattenedVersions",
+                existing,
+              ) ?? [];
+
+            const idx = existingVersions.findIndex((v) => {
+              const versionOrder = readField<number>("order", v) ?? 0;
+              if (minOrder) {
+                return versionOrder - 1 === minOrder;
+              }
+              if (maxOrder) {
+                return versionOrder + 1 === maxOrder;
+              }
+              return false;
+            });
+
+            if (idx === -1) {
+              return undefined;
+            }
+
+            let startIndex = maxOrder ? idx : 0;
+            let endIndex = maxOrder ? existingVersions.length : idx;
+            let numActivated = 0;
+
+            // Count backwards for paginating backwards.
+            if (minOrder) {
```

Comment:
something kind of funky about the cache is that if you have some versions that match the `maxOrder`/`minOrder` conditions (but less than limit), it'll just read from the versions in the cache and not fetch more versions

not sure if this would have to be addressed by [DEVPROD-11707](https://jira.mongodb.org/browse/DEVPROD-11707) or if it's just too difficult to fix 😞

https://github.com/user-attachments/assets/1515985c-219b-4d08-8af3-03c2d040120e

Reply:
Thanks, good catch. When this happens we just want to initiate a new request via returning undefined; better to fetch eagerly than have weirdness. I added handling and some tests.